extensions/parquet/table/src/main/java/io/deephaven/parquet/table/util/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,273 @@
+            currentFragment = finalFragment.asReadOnlyBuffer();
+        } else {
+            // Send async read request the current fragment, if it's not already in the cache
+            final S3ChannelContext.FragmentContext fragmentContext = context.getFragmentContext(currFragmentIndex);
+            final CompletableFuture<ByteBuffer> fetchCurrFragment;
+            if (fragmentContext != null && fragmentContext.fragmentIndex == currFragmentIndex) {
+                fetchCurrFragment = fragmentContext.future;
+            } else {
+                fetchCurrFragment = computeFragmentFuture(currFragmentIndex);
+                context.setFragmentContext(currFragmentIndex, fetchCurrFragment);
+            }
+
+            // Send async requests for read-ahead buffers and store them in the cache
+            final int numFragmentsToLoad = Math.min(readAheadCount, numFragmentsInObject - currFragmentIndex - 1);
+            for (int i = 0; i < numFragmentsToLoad; i++) {
+                final int readAheadFragmentIndex = i + currFragmentIndex + 1;
+                final S3ChannelContext.FragmentContext readAheadFragmentContext = context.getFragmentContext(readAheadFragmentIndex);
+                if (readAheadFragmentContext == null || readAheadFragmentContext.fragmentIndex != readAheadFragmentIndex) {
+                    context.setFragmentContext(readAheadFragmentIndex, computeFragmentFuture(readAheadFragmentIndex));
+                }
+            }
+
+            // Wait till the current fragment is fetched
+            try {
+                currentFragment = fetchCurrFragment.get(timeout, timeUnit).asReadOnlyBuffer();
+            } catch (final InterruptedException | ExecutionException | TimeoutException | RuntimeException e) {
+                throw new UncheckedDeephavenException("Failed to fetch fragment " + currFragmentIndex + " for file "
+                        + key + " in S3 bucket " + bucket, e);
+            }
+            if (currFragmentIndex == finalFragmentIdx) {
+                finalFragment = currentFragment.asReadOnlyBuffer();
+            }
+        }
+
+        // Copy the bytes from fragment from the offset up to the min of remaining fragment and destination bytes.
+        // Therefore, the number of bytes read by this method can be less than the number of bytes remaining in the
+        // destination buffer.
+        final int fragmentOffset = (int) (localPosition - (currFragmentIndex * fragmentSize));
+        currentFragment.position(fragmentOffset);
+        final int limit = Math.min(currentFragment.remaining(), destination.remaining());
+        currentFragment.limit(currentFragment.position() + limit);
+        destination.put(currentFragment);
+        position = localPosition + limit;
+        return limit;
+    }
+
+    private int fragmentIndexForByteNumber(final long byteNumber) {
+        return Math.toIntExact(Math.floorDiv(byteNumber, (long) fragmentSize));
```

Comment:
`floorDiv` is only necessary if we think there will be negative numbers; in this case, I think `Math.toIntExact(byteNumber / fragmentSize)` is appropriate.

For now, I think `int` is appropriate for the fragment index, but if we find the need, it could technically be expanded to a `long`. (Not worth the generalization right now, I wouldn't think - but might leave a comment?)